### PR TITLE
Added P&C lexical audio model

### PR DESCRIPTION
**Pull request by jubick1337**
# What does this PR do ?

Add audio and lexical model to NeMo and necessary support.

**Collection**:  nlp
# Changelog 
- Added `PunctuationCapitalizationLexicalAudioModel` which is inherited from `PunctuationCapitalizationModel`.
- Extended `BertPunctuationCapitalizationDataset` and `BertPunctuationCapitalizationTarredDataset` to support audio as well.
- Added default configs stored in `examples/nlp/token_classification/conf/punctuation_capitalization_lexical_audio_config.yaml`.
- Added script to run train and evaluate at `examples/nlp/token_classification/punctuation_capitalization_lexical_audio_train_evaluate.py`.

# Usage
```
python punctuation_capitalization_lexical_audio_train_evaluate.py \
        model.train_ds.ds_item=<PATH/TO/TRAIN/DATA> \
        model.train_ds.text_file=<NAME_OF_TRAIN_INPUT_TEXT_FILE> \
        model.train_ds.labels_file=<NAME_OF_TRAIN_LABELS_FILE> \
        model.validation_ds.ds_item=<PATH/TO/DEV/DATA> \
        model.validation_ds.text_file=<NAME_OF_DEV_INPUT_TEXT_FILE> \
        model.validation_ds.labels_file=<NAME_OF_DEV_LABELS_FILE
```

# Before your PR is "Ready for review"
**Pre checks**:
- [X] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [X] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:
- [X] New Feature
- [ ] Bugfix
- [X] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed. 
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information
* Related to # (issue)

### CODE REVIEW

File: nemo/collections/nlp/models/token_classification/punctuation_capitalization_config.py [hunk 1]
```diff
@@ -212,6 +213,49 @@ class PunctuationCapitalizationModelConfig:
...
+
+    restore_lexical_encoder_from: Optional[str] = None
+    """"Path to .nemo checkpoint to load weights from"""
+
+    use_adapters: Optional[bool] = False
+    """use adapters for audio encoder"""
+
+    adapter_config: Optional[LinearAdapterConfig] = None
+
+    use_weighted_loss: Optional[bool] = False
```
Comment (ekmb) [inline nemo/collections/nlp/models/token_classification/punctuation_capitalization_config.py:250]: could you remove redundant empty lines to improve the readability so that all related args are grouped together with a comment above the group? e.g., group `use_adapters` and `adapter_config`

File: nemo/collections/nlp/models/token_classification/punctuation_capitalization_config.py [hunk 2]
```diff
@@ -212,6 +213,49 @@ class PunctuationCapitalizationModelConfig:
...
+
+    frozen_conf_d_model: Optional[int] = 256
+
+    frozen_conf_d_ff: Optional[int] = 1024
+
+    frozen_conf_num_layers: Optional[int] = 8
```
Comment (ekmb) [inline nemo/collections/nlp/models/token_classification/punctuation_capitalization_config.py:256]: what is this related to? no adapter frozen conformer?

File: nemo/collections/nlp/models/token_classification/punctuation_capitalization_config.py [hunk 3]
```diff
@@ -212,6 +213,49 @@ class PunctuationCapitalizationModelConfig:
     documentation and `primer <https://github.com/NVIDIA/NeMo/blob/main/tutorials/00_NeMo_Primer.ipynb>_ tutorial."""
 
 
+@dataclass
+class PunctuationCapitalizationLexicalAudioModelConfig(PunctuationCapitalizationModelConfig):
+    train_ds: Optional[PunctuationCapitalizationTrainDataConfig] = None
```
Comment (ekmb) [inline nemo/collections/nlp/models/token_classification/punctuation_capitalization_config.py:218]: add a comment on adapter vs fine-tuning/frozen asr model setup

File: nemo/collections/nlp/models/token_classification/punctuation_capitalization_lexical_audio_model.py [hunk 4]
```diff
@@ -0,0 +1,213 @@
...
+        ``attention_mask`` arguments). See more about model config options :ref:`here<model-config-label>`.
+        And any :class:`~nemo.collections.asr.models.EncDecCTCModel` which has encoder module which is used as an AM.
+
+        For training and testing use dataset
+        :class:`~nemo.collections.nlp.data.token_classification.punctuation_capitalization_dataset.BertPunctuationCapitalizationDataset` with parameter ``use_audio`` set to ``True``,
+        for training on huge amounts of data which cannot be loaded into memory simultaneously use
+        :class:`~nemo.collections.nlp.data.token_classification.punctuation_capitalization_tarred_dataset.BertPunctuationCapitalizationTarredDataset` with parameter ``use_audio`` set to ``True``.
+
+        Args:
+            cfg (:obj:`DictConfig`): a model configuration. It should follow dataclass
```
Comment (ekmb) [inline nemo/collections/nlp/models/token_classification/punctuation_capitalization_lexical_audio_model.py:60]: remove type, it's already in line 68

File: nemo/collections/nlp/models/token_classification/punctuation_capitalization_lexical_audio_model.py [hunk 5]
```diff
@@ -0,0 +1,213 @@
...
+            self.audio_encoder.cfg.encoder.d_model, self.bert_model(**self.bert_model.input_example()[0]).size()[-1]
+        )
+
+        if cfg.get('freeze_audio_encoder', False):
+            for param in self.audio_encoder.parameters():
+                param.requires_grad = False
+            for i in range(cfg.get('frozen_conf_num_layers')):
+                self.audio_encoder.add_module(
+                    f'conf_encoder_{i}',
+                    ConformerLayer(d_model=cfg.get('frozen_conf_d_model'), d_ff=cfg.get('frozen_conf_d_ff')),
```
Comment (ekmb) [inline nemo/collections/nlp/models/token_classification/punctuation_capitalization_lexical_audio_model.py:102]: Do the model work only with Conformer? no citrinet?

File: nemo/collections/nlp/data/token_classification/punctuation_capitalization_dataset.py [hunk 6]
```diff
@@ -477,8 +527,15 @@ def __call__(
...
         all_input_ids, all_subtokens_mask, punct_all_labels, capit_all_labels = [], [], [], []
+        dummy = [
+            None for _ in range(len(queries))
+        ]  # Needed to avoid code duplication with different values of `self.use_audio`
+        all_audio_waveforms = [] if preload_audios else dummy
+        audio_lengths = [] if preload_audios else dummy
+        audio_filepaths = [] if not preload_audios else dummy
         progress_made = 0
-        for i, query in enumerate(queries):
+        queries = zip(queries, audio_queries) if audio_queries else zip(queries, [None for _ in range(len(queries))])
```
Comment (ekmb) [inline nemo/collections/nlp/data/token_classification/punctuation_capitalization_dataset.py:537]: second arg in the last zip the `dummy` variable.

**Review by ekmb — CHANGES_REQUESTED** [review]
Could you fix LGTM errors?

File: examples/nlp/token_classification/conf/punctuation_capitalization_lexical_audio_config.yaml [hunk 7]
```diff
@@ -0,0 +1,187 @@
...
+    mode: "min"
+    save_best_model: true
+
+model:
+  pretrained_audio_encoder: stt_en_conformer_ctc_large # You can choose any pretrained ASR model from list_available_models() of EncDecCTCModel.
+
+  freeze_audio_encoder: false # If set to True weights of audio encoder will not be updated during training.
+
+  use_adapters: false # If set to True will enable adapters for audio encoder. Also, should provide config of adapter.
+
```
[the remark this answers is not in the document]
Comment (jubick1337) [reply examples/nlp/token_classification/conf/punctuation_capitalization_lexical_audio_config.yaml:55]: > how about having:
> 
> ```
> audio_encoder:
>      name: stt_en_conformer_ctc_large
>      freeze: false
>      use_adapters: false
> ```
> 
> So that everything related to autoencoder in the same config field

This change probably would break all of current checkpoints as it requires changes of config structure

File: nemo/collections/nlp/data/token_classification/punctuation_capitalization_tarred_dataset.py [hunk 8]
```diff
@@ -113,19 +114,34 @@ def get_fragment_start_bytes(
...
-    result = Parallel(n_jobs=2)(
-        delayed(count_lines_and_get_fragment_starting_positions)(file_name, lines_per_dataset_fragment)
-        for file_name in [text_file, labels_file]
-    )
-    if result[0][0] != result[1][0]:
-        raise ValueError(
-            f"Text file {text_file} and label file {labels_file} contain different number of lines. Number of lines "
-            f"in text file: {result[0][0]}, number of lines in label file: {result[1][0]}."
+    if audio_file:
+        result = Parallel(n_jobs=3)(
```
[the remark this answers is not in the document]
Comment (jubick1337) [reply nemo/collections/nlp/data/token_classification/punctuation_capitalization_tarred_dataset.py:118]: Because in this case we need to process 3 files simultaneously, below it's 2 files

**Review by ekmb — CHANGES_REQUESTED** [review]
please create a new environment and install only NLP collection to check that no additional ASR import guard is needed

File: examples/nlp/token_classification/punctuation_capitalization_lexical_audio_train_evaluate.py [hunk 9]
```diff
@@ -0,0 +1,155 @@
...
+            class_labels=cfg.model.class_labels,
+            common_dataset_parameters=cfg.model.common_dataset_parameters,
+            train_ds=cfg.model.get('train_ds') if cfg.do_training else None,
+            validation_ds=cfg.model.get('validation_ds') if cfg.do_training else None,
+            test_ds=cfg.model.get('test_ds') if cfg.do_testing else None,
+            optim=cfg.model.get('optim') if cfg.do_training else None,
+        )
+        model.set_trainer(trainer)
+        if cfg.do_training:
+            model.setup_training_data()
```
[the remark this answers is not in the document]
Comment (jubick1337) [reply examples/nlp/token_classification/punctuation_capitalization_lexical_audio_train_evaluate.py:143]: I don't think such thing exists

File: nemo/collections/nlp/models/token_classification/punctuation_capitalization_lexical_audio_model.py [hunk 10]
```diff
@@ -0,0 +1,260 @@
...
+            )
+            capit_weight = 1 - (capit_freq - capit_freq.min()) / capit_freq.max()
+
+            self.loss_punct = CrossEntropyLoss(logits_ndim=3, weight=punct_weight)
+            self.loss_capit = CrossEntropyLoss(logits_ndim=3, weight=capit_weight)
+        else:
+            self.loss_punct = self.loss
+            self.loss_capit = self.loss
+
+        self.set_max_audio_length(1024)
```
[the remark this answers is not in the document]
Comment (jubick1337) [reply nemo/collections/nlp/models/token_classification/punctuation_capitalization_lexical_audio_model.py:144]: Because it does not matter in this case. It's going to be updated at first call of `forward` and not used before.
It makes sense in ASR config but not here as it's only purpose is keep track of max length for mask creation

**Review by ekmb — APPROVED** [review]
